pennylane/ops/qubit/observables.py

```diff
@@ -438,11 +438,16 @@ class BasisStateProjector(Projector, Operation):
     # The call signature should be the same as Projector.__new__ for the positional
     # arguments, but with free key word arguments.
     def __init__(self, state, wires, id=None):
-        wires = Wires(wires)
-        state = list(qml.math.toarray(state).astype(int))
+        wires = qml.wires.Wires(wires)
 
-        if not set(state).issubset({0, 1}):
-            raise ValueError(f"Basis state must only consist of 0s and 1s; got {state}")
+        if not qml.math.is_abstract(state):
+            # for cases like [jax.numpy.array(0), jax.numpy.array(1)]
+            state_array = qml.math.asarray(state)
+            state_array = qml.math.cast(state_array, int)
+            if not all(x in [0, 1] for x in state_array):
+                raise ValueError(f"Basis state must only consist of 0s and 1s; got {state_array}")
+        else:
+            state_array = qml.math.asarray(state)
 
         super().__init__(state, wires=wires, id=id)
```

Comment:
The problems with the generator of a controlled gate should be able to be resolved by returning `state_array` here instead of `state`.  


Basically the generator of a controlled was providing an array of booleans to `Projector` (the control values), and the booleans were messing up the matrix calculation.